nvflare/client/in_process/api.py

```diff
@@ -0,0 +1,203 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+import time
+from typing import Any, Dict, Optional, Union
+
+from nvflare.apis.analytix import AnalyticsDataType
+from nvflare.apis.fl_constant import FLMetaKey
+from nvflare.apis.shareable import Shareable
+from nvflare.app_common.abstract.fl_model import FLModel, ParamsType
+from nvflare.app_common.utils.fl_model_utils import FLModelUtils
+from nvflare.client.api_spec import APISpec
+from nvflare.client.config import ClientConfig, ConfigKey, from_file
+from nvflare.client.constants import SYS_ATTRS
+from nvflare.client.utils import DIFF_FUNCS
+from nvflare.fuel.data_event.data_bus import DataBus
+from nvflare.fuel.data_event.event_manager import EventManager
+
+TOPIC_LOG_DATA = "LOG_DATA"
+TOPIC_STOP = "STOP"
+TOPIC_ABORT = "ABORT"
+TOPIC_LOCAL_RESULT = "LOCAL_RESULT"
+TOPIC_GLOBAL_RESULT = "GLOBAL_RESULT"
+
+
+class InProcessClientAPI(APISpec):
+    def __init__(self, job_metadata: dict, result_check_interval: float = 2.0):
+        self.data_bus = DataBus()
+        self.data_bus.subscribe([TOPIC_GLOBAL_RESULT], self.__receive_callback)
+        self.data_bus.subscribe([TOPIC_ABORT, TOPIC_STOP], self.__ask_to_abort)
+
+        self.meta = job_metadata
+        self.result_check_interval = result_check_interval
+
+        self.start_round = None
+        self.fl_model = None
+        self.sys_info = {}
+        self.client_config: Optional[ClientConfig] = None
+        self.current_round = None
+        self.total_rounds = None
+        self.logger = logging.getLogger(self.__class__.__name__)
+        self.event_manager = EventManager(self.data_bus)
+        self.abort_reason = ""
+        self.stop_reason = ""
+        self.abort = False
+        self.stop = False
+
+    def init(self, config: Union[str, Dict] = None, rank: Optional[str] = None):
+        config = {} if config is None else config
+        self.prepare_client_config(config)
+
+        for k, v in self.client_config.config.items():
+            if k in SYS_ATTRS:
+                self.sys_info[k] = v
+
+    def prepare_client_config(self, config):
+        if isinstance(config, str):
+            client_config = from_file(config_file=config)
+        elif isinstance(config, dict):
+            client_config = ClientConfig(config=config)
+        else:
+            raise ValueError("config should be either a string or dictionary.")
+
+        if client_config.config:
+            client_config.config.update(self.meta)
+        else:
+            client_config.config = self.meta
+        self.client_config = client_config
```

Comment:
ex_process/api.py takes in a str for a file name since it uses the external_configurator.py to write a config file to "client_api_config.json". For the in_process/api.py case do we need to support an external config file?

Reply:
Update: remove "str" option from `in_process` since there is no need for external config file. Keep adding meta in `prepare_client_config()` so there is still option to add config in `init()`.